### PR TITLE
8759 financing validation by reg type.

### PR DESCRIPTION
**Pull request by doug-lovett**
Signed-off-by: Doug Lovett <doug@diamante.ca>

*Issue #:* /bcgov/entity#8759

*Description of changes:*
Apply data validation rules in the PPR Registration Types spreadsheet to the API create financing statement request.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

**Review by kialj876 — APPROVED** [review]
Looks good 👍  Should fix the linter before merging though

File: ppr-api/src/ppr_api/utils/validators/financing_validator.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        return error_msg
+
+
+def validate_life(json_data, reg_type: str, reg_class: str):
+    """Validate lifeYears and lifeInfinite by registration type."""
+    error_msg = ''
+    if reg_type in LIFE_INFINITE_LIST or reg_class in (model_utils.REG_CLASS_CROWN, model_utils.REG_CLASS_MISC):
+        if 'lifeYears' in json_data and (json_data['lifeYears'] != model_utils.LIFE_INFINITE):
+            error_msg = LY_NOT_ALLOWED
+        if 'lifeInfinite' in json_data and not json_data['lifeInfinite']:
```
Comment (kialj876) [inline ppr-api/src/ppr_api/utils/validators/financing_validator.py:98]: I noticed you have a lot of checks for the field and then another check for the value if the field is there. Might be simpler for you to use the *json_data.get('field', None) !/==* notation for these in the future?

Comment (doug-lovett) [reply ppr-api/src/ppr_api/utils/validators/financing_validator.py:98]: Good idea.

File: ppr-api/src/ppr_api/models/registration.py [hunk 2]
```diff
@@ -36,6 +36,53 @@
 FINANCING_PATH = '/ppr/api/v1/financing-statements/'
 
 
+class CrownChargeTypes(Enum):
```
Comment (thorwolpert) [inline ppr-api/src/ppr_api/models/registration.py]:
```suggestion
class CrownChargeTypes(str, Enum):
```
makes it comparable without having to to do ```enum.value```

File: ppr-api/src/ppr_api/utils/validators/financing_validator.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            error_msg += LI_INVALID
+    elif reg_type == model_utils.REG_TYPE_REPAIRER_LIEN and 'lifeInfinite' in json_data and json_data['lifeInfinite']:
+        error_msg += LI_NOT_ALLOWED
+    elif 'lifeYears' not in json_data and 'lifeInfinite' not in json_data:
+        error_msg += LIFE_MISSING
+    elif 'lifeYears' in json_data and 'lifeInfinite' in json_data and \
```
Comment (thorwolpert) [inline ppr-api/src/ppr_api/utils/validators/financing_validator.py]:
```suggestion
    elif json_data.get('lifeInfinite') and json_data.get('lifeYears', -1) > 0
```
wont throw exceptions for missing values

File: ppr-api/src/ppr_api/utils/validators/financing_validator.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            error_msg += RL_DATE_INVALID
+
+    return error_msg
+
+
+def get_registration_class(reg_type: str):
+    """Derive the registration class from the registration type."""
+    try:
+        reg_class = model_utils.REG_TYPE_TO_REG_CLASS[reg_type]
+        if reg_class in (model_utils.REG_CLASS_CROWN, model_utils.REG_CLASS_MISC, model_utils.REG_CLASS_PPSA):
```
Comment (thorwolpert) [inline ppr-api/src/ppr_api/utils/validators/financing_validator.py]:
```suggestion
        if (reg_class := model_utils.REG_TYPE_TO_REG_CLASS[reg_type]) in \
        (model_utils.REG_CLASS_CROWN, model_utils.REG_CLASS_MISC, model_utils.REG_CLASS_PPSA):
```
the walrus is your friend :=

**Review by thorwolpert — COMMENTED** [review]
a couple of suggestions

File: ppr-api/src/ppr_api/utils/validators/financing_validator.py (hunk 4 above)
Comment (doug-lovett) [reply ppr-api/src/ppr_api/utils/validators/financing_validator.py]: Suggested change made.

File: ppr-api/src/ppr_api/utils/validators/financing_validator.py (hunk 3 above)
Comment (doug-lovett) [reply ppr-api/src/ppr_api/utils/validators/financing_validator.py]: Suggested change made.

File: ppr-api/src/ppr_api/models/registration.py (hunk 2 above)
Comment (doug-lovett) [reply ppr-api/src/ppr_api/models/registration.py]: Suggested change made.

**Pull request merged**